File: Loenn/entities/dash_through_spikes.lua
local spikeHelper = require("helpers.spikes")

local handlers = $("Up", "Down", "Left", "Right"):map((dir -> spikeHelper.createEntityHandler("NerdHelper/DashThroughSpikes" .. dir, dir:lower(), false, false, {"Kalobi/NerdHelper/dashthroughspike"})))
for k, handler in lpairs(handlers) do
    handler.placements[1].name = "normal_spike"
    handler.placements[1].data = ($(handler.placements[1].data) .. $({red_boosters_count_as_dash = true, invert = false, zero_speed_only = false, along = true, into = true, diag = true}))()
    handler.placements[2] = {name = "fixer", data = table.shallowcopy(handler.placements[1].data)}
    handler.placements[2].data.type = "default"
    handler.placements[2].data.zero_speed_only = true
    handler.fieldInformation["type"].options = table.shallowcopy(spikeHelper.spikeVariants)
    table.insert(handler.fieldInformation["type"].options, "Kalobi/NerdHelper/dashthroughspike")
    handler.fieldInformation["type"].editable = true
    handler.fieldOrder = {"x", "y", (handler.placements[1].data["height"] ? "height" : "width"), "type", "red_boosters_count_as_dash", "along", "into", "diag", "zero_speed_only", "invert"}
end
return handlers()

File: Code/Entities/DashThroughSpikes.cs
using System;
using Celeste.Mod.Entities;
using Celeste.Mod.NerdHelper.Module;
using Microsoft.Xna.Framework;
using Monocle;

namespace Celeste.Mod.NerdHelper.Entities;

[CustomEntity(
    "NerdHelper/DashThroughSpikesUp = LoadUp",
    "NerdHelper/DashThroughSpikesDown = LoadDown",
    "NerdHelper/DashThroughSpikesLeft = LoadLeft",
    "NerdHelper/DashThroughSpikesRight = LoadRight"
)]
[TrackedAs(typeof(Spikes))]
public class DashThroughSpikes : Spikes {
    private const string defaultTexture = "Kalobi/NerdHelper/dashthroughspike";
    
    private readonly Action<Player> origOnCollide;

    private void OnPlayer(Player player) {
        if (((player.StateMachine.State == Player.StDash
              || player.DashAttacking && player.StateMachine.State != Player.StRedDash
              || player.StateMachine.State == Player.StDreamDash
              || (NerdHelperModule.CommunalHelperInterop.GetDreamTunnelDashState != null
                  && player.StateMachine.State == NerdHelperModule.CommunalHelperInterop.GetDreamTunnelDashState())
              || (NerdHelperModule.CommunalHelperInterop.IsSeekerDashAttacking?.Invoke() ?? false)
              || player.StateMachine.State == Player.StRedDash && red)
             && (!zeroSpeedOnly || player.Speed.Equals(Vector2.Zero))
             && CheckDir(player.DashDir))
            ^ invert) {
            return;
        }
        origOnCollide(player);
    }

    private bool CheckDir(Vector2 dashDir) {
        if (Direction is Directions.Up or Directions.Down) {
            return dashDir.X != 0 && dashDir.Y == 0 && along
                   || dashDir.X == 0 && dashDir.Y != 0 && into
                   || dashDir.X != 0 && dashDir.Y != 0 && diag
                   || dashDir.Equals(Vector2.Zero);
        }
        return dashDir.X != 0 && dashDir.Y == 0 && into
               || dashDir.X == 0 && dashDir.Y != 0 && along
               || dashDir.X != 0 && dashDir.Y != 0 && diag
               || dashDir.Equals(Vector2.Zero);
    }

    private bool red;
    private bool invert;
    private bool zeroSpeedOnly;
    private bool into;
    private bool along;
    private bool diag;

    public static Entity LoadUp(Level level, LevelData levelData, Vector2 offset, EntityData entityData) {
        return new DashThroughSpikes(entityData, offset, Directions.Up);
    }

    public static Entity LoadDown(Level level, LevelData levelData, Vector2 offset, EntityData entityData) {
        return new DashThroughSpikes(entityData, offset, Directions.Down);
    }

    public static Entity LoadLeft(Level level, LevelData levelData, Vector2 offset, EntityData entityData) {
        return new DashThroughSpikes(entityData, offset, Directions.Left);
    }

    public static Entity LoadRight(Level level, LevelData levelData, Vector2 offset, EntityData entityData) {
        return new DashThroughSpikes(entityData, offset, Directions.Right);
    }

    public DashThroughSpikes(EntityData data, Vector2 offset, Directions dir) : base(data, offset, dir) {
        red = data.Bool("red_boosters_count_as_dash", true);
        invert = data.Bool("invert", false);
        zeroSpeedOnly = data.Bool("zero_speed_only", false);
        into = data.Bool("into", true);
        along = data.Bool("along", true);
        diag = data.Bool("diag", true);
        
        string texturePath = data.Attr("type", defaultTexture);
        if (texturePath.Length == 0) {
            texturePath = defaultTexture;
        }
        overrideType = texturePath;
        
        origOnCollide = pc.OnCollide;
        pc.OnCollide = OnPlayer;
    }
}